// react-final-form-graceful-field/GracefulField.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var React = _interopRequireWildcard(require("react"));

var _useGracefulField = _interopRequireDefault(require("./useGracefulField"));

var GracefulField = React.forwardRef(function GracefulField(_ref, ref) {
  var afterSubmit = _ref.afterSubmit,
      allowNull = _ref.allowNull,
      beforeSubmit = _ref.beforeSubmit,
      children = _ref.children,
      component = _ref.component,
      render = _ref.render,
      data = _ref.data,
      defaultValue = _ref.defaultValue,
      format = _ref.format,
      initialValue = _ref.initialValue,
      invalidValue = _ref.invalidValue,
      isEqual = _ref.isEqual,
      multiple = _ref.multiple,
      name = _ref.name,
      parse = _ref.parse,
      subscription = _ref.subscription,
      type = _ref.type,
      validate = _ref.validate,
      validateFields = _ref.validateFields,
      value = _ref.value,
      rest = (0, _objectWithoutProperties2.default)(_ref, ["afterSubmit", "allowNull", "beforeSubmit", "children", "component", "render", "data", "defaultValue", "format", "initialValue", "invalidValue", "isEqual", "multiple", "name", "parse", "subscription", "type", "validate", "validateFields", "value"]);

  if (!name) {
    throw new Error('prop name cannot be undefined in <GracefulField> component');
  }

  var field = (0, _useGracefulField.default)(name, {
    afterSubmit: afterSubmit,
    allowNull: allowNull,
    beforeSubmit: beforeSubmit,
    children: children,
    component: component,
    data: data,
    defaultValue: defaultValue,
    format: format,
    initialValue: initialValue,
    isEqual: isEqual,
    multiple: multiple,
    parse: parse,
    subscription: subscription,
    type: type,
    validate: validate,
    validateFields: validateFields,
    value: value
  });

  if (typeof children === 'function') {
    return children((0, _objectSpread2.default)({}, field, rest));
  }

  if (typeof component === 'string') {
    // ignore meta, combine input with any other props
    // eslint-disable-next-line react/no-children-prop
    return React.createElement(component, (0, _objectSpread2.default)({}, field.input, {
      children: children,
      ref: ref
    }, rest));
  }

  if (component) {
    return React.createElement(component, (0, _objectSpread2.default)({
      ref: ref
    }, rest, field));
  }

  if (typeof render !== 'function') {
    throw new Error("Must specify either a render prop, a render function as children, or a component prop to GracefulField(".concat(name, ")"));
  }

  return render(children === undefined ? (0, _objectSpread2.default)({
    ref: ref
  }, rest, field) : (0, _objectSpread2.default)({
    ref: ref
  }, rest, field, {
    children: children
  }));
});
var _default = GracefulField;
exports.default = _default;
